### ISSUE TITLE
twisted.web._newclient.Response leaves its transport in an unpredictable state depending on how large the response body is

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6751]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#6751|
|Type|defect|
|Created|2013-09-18 04:26:44Z|
|Branch|https://github.com/twisted/twisted/tree/deliverBody-hang-6751|

If an HTTP response has a small response-body, and a Twisted application does not immediately call `deliverBody` when said response is received, `Response` can leave its transport paused.

When I say "immediately", I mean, _synchronously within the response callback's invocation_.  If you call `deliverBody` at any point later, you go through the `_deliverBody_DEFERRED_CLOSE` callback which doesn't work.

Among other things, this means that when `HTTP11ClientProtocol.quiescentCallback` is invoked, the 'quiescent' connection might be reading (if it happened to go through `_deliverBody_INITIAL`) or it might be paused (if it happened to go through `_deliverBody_DEFERRED_CLOSE`).

This will lead to two negative effects:

 1. `twisted.web.client.HTTPConnectionPool` will end up adding paused connections to its persistent connection cache, and
 2. if something manages to be holding on to the transport, it'll silently hang around forever, leaking an FD; if nothing is holding on to it, it'll be GC'd, closing its socket (without calling connectionLost) some time later.

<details><summary>Searchable metadata</summary>

```
trac-id__6751 6751
type__defect defect
reporter__glyph glyph
priority__high high
milestone__ 
branch__branches_deliverBody_hang_6751 branches/deliverBody-hang-6751
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1379478404000000 1379478404000000
changetime__1390870798000000 1390870798000000
version__None None
owner__glyph glyph
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set status to `closed`|
|-|-|

Duplicate of [#6711](https://github.com/twisted/twisted/issues/6711) (but maybe you want to move the description over there, since it's better).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set status to `reopened`|
|-|-|

Incorrect.  This occurs even if you _do_ call `deliverBody`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

(In [41416]) Branching to deliverBody-hang-6751

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set status to `new`|
|-|-|

I think the code in the branch should address this.  Setting the priority to "high" since this is actually a pretty serious resource leak that can be very confusing to debug.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

What happens when you never call `deliverBody`? Feel free to say "I opened a different ticket for that", but maybe you could solve it at same time, who knows.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_8):
> What happens when you never call `deliverBody`? Feel free to say "I opened a different ticket for that", but maybe you could solve it at same time, who knows.

So, if you never call `deliverBody`, the connection never gets added back to the pool.

If the server wants to time you out, it will time you out and the connection will be removed from the pool (and the reactor, and you'll disconnect).

If the server wants to let the connection hang out forever, it can hang out forever.

In other words, now not calling `deliverBody` is, in a sense, a perfectly valid thing to do, and you'll get exactly the behavior that one would expect from that.

As it happens, never calling `deliverBody` _won't_ hang the connection pool, because the connection pool just makes sure that there are no more than `maxPersistentPerHost` _idle_ connections.

Arguably, there should be a way to set a default timeout on how long one should take before calling `deliverBody`, but that is more of a new feature than a bug fix.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @wsanchez|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_12"></a>|@wsanchez set owner to @glyph<br>@wsanchez set status to `new`|
|-|-|

Reviewed.  No suggested changes.

Kicked the Windows builder that ran out of disk space, passed this time.  WinXP failed due in unrelated stuff because it's Windows or something.

OK to merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

Replying to [wsanchez](#note_12):
> Reviewed.  No suggested changes.

Thanks a bunch!

> Kicked the Windows builder that ran out of disk space, passed this time.  WinXP failed due in unrelated stuff because it's Windows or something.

FYI, the Windows builder isn't actually running out of disk space.  It's running out of "being a functioning operating system".  We could fix this bug in Twisted and man we really should so this stops being a perennial comment in reviews.  [ticket:5921 You can see the ticket describing the problem, which even has a branch already, although is not in review, here.]

> OK to merge.

OK, will land soon.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set status to `closed`|
|-|-|

(In [41439]) Merge deliverBody-hang-6751: Fix a hang in Agent when using a connection pool.

Author: glyph

Reviewer: wsanchez

Fixes: [#6751](https://github.com/twisted/twisted/issues/6751)

twisted.web.client.Agent now correctly manage flow-control on pooled
connections, and therefore requests will no longer hang sometimes when
deliverBody is not called synchronously within the callback on Request.
